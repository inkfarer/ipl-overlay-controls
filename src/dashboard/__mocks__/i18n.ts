import i18next from 'i18next';
import { config } from '@vue/test-utils';

jest.mock('i18next-vue', () => ({
    __esModule: true,
    useTranslation: () => ({ t: i18next.t })
}));

beforeAll(async () => {
    await i18next.init({
        lng: 'cimode',
        appendNamespaceToCIMode: true
    });
    config.global.mocks = {
        $t: i18next.t
    };
});
