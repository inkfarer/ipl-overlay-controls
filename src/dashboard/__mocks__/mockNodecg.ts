import Mock = jest.Mock;

export type ReplicantChangeHandler = (newValue?: unknown, oldValue?: unknown) => void;
export type MockReplicant = { value: unknown };
type OnFunction = (event: string, handler: ReplicantChangeHandler) => void;

export class MockNodecg {
    replicants: {[key: string]: MockReplicant};
    listeners: {[key: string]: ReplicantChangeHandler};
    sendMessage: Mock

    constructor() {
        this.replicants = {};
        this.listeners = {};
        this.sendMessage = jest.fn();
    }

    init(): void {
        const self = this;
        window.nodecg = {
            // @ts-ignore: Just a mock
            Replicant(name: string) {
                const replicantValue: { value: unknown, on: OnFunction } = {
                    value: undefined,
                    on: (event: string, handler: ReplicantChangeHandler) => {
                        self.listeners[name] = handler;
                    }
                };
                self.replicants[name] = replicantValue;
                return replicantValue;
            },
            sendMessage: self.sendMessage,
            // @ts-ignore: Just a mock
            log: {
                info: jest.fn()
            }
        };

        // @ts-ignore: It's a mock.
        window.NodeCG = {
            waitForReplicants: jest.fn().mockResolvedValue(true)
        };
    }
}
