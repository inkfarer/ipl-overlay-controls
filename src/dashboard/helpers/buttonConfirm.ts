/*
buttonConfirm

After importing this script to a dashboard panel, add the attribute 'data-requires-confirmation' to any button.
The button will now ask for confirmation when it is first clicked.
When it is clicked a second time, a 'confirm' event is sent by it, which can be listened for by scripts.

Example:

HTML:
    <button id="dangerous-action-button" data-requires-confirmation>DO A THING</button>
JS:
    document.getElementById('dangerous-action-button').addEventListener('confirm', () => {
        // do something
    });
*/

window.addEventListener('DOMContentLoaded', () => {
    function addListener(btn: HTMLButtonElement) {
        if (btn.dataset.hasConfirmListener !== 'true') {
            btn.addEventListener('click', e => {
                const target = e.target as HTMLButtonElement;

                function resetButton() {
                    target.dataset.isClicked = 'false';
                    target.innerText = target.dataset.originalText;
                    target.style.backgroundColor = '';
                }

                if (target.dataset.isClicked === 'true') {
                    target.dispatchEvent(new Event('confirm'));
                    resetButton();
                } else {
                    target.dataset.isClicked = 'true';
                    target.dataset.originalText = target.innerText;
                    target.style.backgroundColor = 'var(--red)';
                    target.innerText = 'are you sure?';

                    setTimeout(() => {
                        resetButton();
                    }, 5000);
                }
            });

            btn.dataset.hasConfirmListener = 'true';
        }
    }

    const buttons = document.querySelectorAll('button[data-requires-confirmation]');
    buttons.forEach(addListener);

    const observer = new MutationObserver(mutations => {
        mutations.forEach(mutation => {
            mutation.addedNodes.forEach(node => {
                if (node.nodeType === Node.ELEMENT_NODE
                    && node.nodeName === 'BUTTON'
                    && (node as HTMLElement).dataset.requiresConfirmation) {

                    const btn = node as HTMLButtonElement;
                    addListener(btn);
                }
            });
        });
    });

    observer.observe(document.body, {
        subtree: true,
        childList: true
    });
});
