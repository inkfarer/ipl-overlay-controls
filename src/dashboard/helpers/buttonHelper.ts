import { colors } from '../styles/colors';

export function setToggleButtonDisabled(
    toggleTrue: HTMLButtonElement,
    toggleFalse: HTMLButtonElement,
    state: boolean
): void {
    toggleTrue.disabled = state;
    toggleFalse.disabled = !state;
}

export function addChangeReminder(
    elements: HTMLElement[] | NodeListOf<HTMLElement>,
    updateButton: HTMLButtonElement
): void {
    elements.forEach(element => {
        if (!element.tagName) return;

        let event;
        if (element.tagName.toLowerCase() === 'input') event = 'input';
        else if (element.tagName.toLowerCase() === 'select') event = 'change';
        else return;

        element.addEventListener(event, e => {
            const target = e.target as HTMLElement;
            target.dataset.edited = 'true';
            updateButton.style.backgroundColor = colors.red;
        });
    });

    updateButton.addEventListener('click', () => {
        elements.forEach(elem => {
            elem.dataset.edited = 'false';
        });
        updateButton.style.backgroundColor = colors.blue;
    });
}
