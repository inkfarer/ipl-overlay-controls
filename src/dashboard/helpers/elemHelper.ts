export function appendChildren(element: HTMLElement, ...children: HTMLElement[]): void {
    children.forEach(elem => {
        element.appendChild(elem);
    });
}

export function elementById<T extends HTMLElement>(id: string): T | undefined {
    return document.getElementById(id) as T;
}

export function dispatch(element: EventTarget, event: string): void {
    element.dispatchEvent(new Event(event));
}

export function hideElement(element: HTMLElement): void {
    element.style.display = 'none';
}

export function showElement(element: HTMLElement): void {
    element.style.removeProperty('display');
}

export function prependToBody(elem: HTMLElement): void {
    document.body.prepend(elem);
}

export function appendElementAfter(newElement: HTMLElement, element: HTMLElement): void {
    element.parentNode.insertBefore(newElement, element.nextSibling);
}
