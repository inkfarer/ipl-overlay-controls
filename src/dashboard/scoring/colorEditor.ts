import { colors } from './colors';
import { scoreboardData } from './replicants';
import { ColorInfo } from 'types/colorInfo';

const customColorToggle = <HTMLInputElement> document.getElementById('custom-color-toggle');
const colorSelector = <HTMLSelectElement> document.getElementById('color-selector');

for (let i = 0; i < colors.length; i++) {
    const element = colors[i];

    const optGroup = document.createElement('optgroup');
    optGroup.label = element.meta.name;

    for (let j = 0; j < element.colors.length; j++) {
        const color = element.colors[j];

        const option = document.createElement('option');
        option.value = `${formatCategoryName(element.meta.name)}_${color.index}`;
        option.text = color.title;
        option.dataset.index = String(color.index);
        option.dataset.firstColor = color.clrA;
        option.dataset.secondColor = color.clrB;
        option.dataset.categoryName = element.meta.name;
        // Make custom color unselectable by user
        option.disabled = color.index === 999;

        optGroup.appendChild(option);
    }

    colorSelector.appendChild(optGroup);
}

scoreboardData.on('change', newValue => {
    colorSelector.value = `${formatCategoryName(newValue.colorInfo.categoryName)}_${newValue.colorInfo.index}`;

    updateColorDisplay(
        (newValue.colorInfo as ColorInfo),
        document.getElementById('team-a-color-display'),
        'a',
        newValue.swapColorOrder
    );
    updateColorDisplay(
        (newValue.colorInfo as ColorInfo),
        document.getElementById('team-b-color-display'),
        'b',
        newValue.swapColorOrder
    );
    updateColorDisplay(
        (newValue.colorInfo as ColorInfo),
        document.getElementById('team-a-custom-color'),
        'a',
        newValue.swapColorOrder
    );
    updateColorDisplay(
        (newValue.colorInfo as ColorInfo),
        document.getElementById('team-b-custom-color'),
        'b',
        newValue.swapColorOrder
    );

    const customColorEnabled = newValue.colorInfo.index === 999;
    customColorToggle.checked = customColorEnabled;
    updateCustomColorToggle(customColorEnabled);
});

function updateColorDisplay(colorInfo: ColorInfo, elem: HTMLElement, team: 'a' | 'b', swapColors: boolean) {
    let color;

    if (team === 'a' && !swapColors) color = colorInfo.clrA;
    else if (team === 'a' && swapColors) color = colorInfo.clrB;
    else if (team === 'b' && !swapColors) color = colorInfo.clrB;
    else if (team === 'b' && swapColors) color = colorInfo.clrA;
    else color = '#000000';

    switch (elem.tagName.toLowerCase()) {
        case 'input':
            (elem as HTMLInputElement).value = color;
            break;
        default:
            elem.style.backgroundColor = color;
    }
}

customColorToggle.onchange = e => {
    updateCustomColorToggle((e.target as HTMLInputElement).checked);
};

function updateCustomColorToggle(checked: boolean) {
    const colorSelectContainer = document.getElementById('color-select-container');
    const customColorContainer = document.getElementById('custom-color-select-container');

    if (checked) {
        colorSelectContainer.style.display = 'none';
        customColorContainer.style.display = 'flex';
    } else {
        colorSelectContainer.style.display = 'unset';
        customColorContainer.style.display = 'none';
    }
}

document.getElementById('swap-color-order-btn').onclick = () => {
    scoreboardData.value.swapColorOrder = !scoreboardData.value.swapColorOrder;
};

document.getElementById('update-scoreboard-btn').addEventListener('click', () => {
    const colorOption = <HTMLOptionElement> colorSelector.options[colorSelector.selectedIndex];

    let clrInfo: ColorInfo;
    let { swapColorOrder } = scoreboardData.value;

    if (customColorToggle.checked) {
        clrInfo = {
            index: 999,
            title: 'Custom Color',
            clrA: (document.getElementById('team-a-custom-color') as HTMLInputElement).value,
            clrB: (document.getElementById('team-b-custom-color') as HTMLInputElement).value,
        };
        swapColorOrder = false;
    } else {
        clrInfo = {
            index: Number(colorOption.dataset.index),
            title: colorOption.text,
            clrA: colorOption.dataset.firstColor,
            clrB: colorOption.dataset.secondColor,
        };
    }

    scoreboardData.value.colorInfo = {
        ...clrInfo,
        categoryName: colorOption.dataset.categoryName
    };
    scoreboardData.value.swapColorOrder = swapColorOrder;
});

function formatCategoryName(name: string): string {
    return name.replace(' ', '-').toLowerCase();
}
