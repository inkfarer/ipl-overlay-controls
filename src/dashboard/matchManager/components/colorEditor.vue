<template>
    <ipl-expanding-space
        key="more-colors"
        title="More colors"
    >
        <div class="colors-container">
            <div class="layout horizontal center-horizontal">
                <ipl-checkbox
                    v-model="useCustomColor"
                    label="Use custom color"
                    small
                />
            </div>
            <div
                v-if="!useCustomColor"
                class="m-t-8"
            >
                <div
                    v-for="(group, groupIndex) in colorsWithoutCustom"
                    :key="`color-group_${groupIndex}`"
                    class="color-group"
                >
                    <div class="title">{{ group.meta.name }}</div>
                    <div
                        v-for="(color, colorIndex) in group.colors"
                        :key="`color_${groupIndex}_${colorIndex}`"
                        :data-test="`color-option-${groupIndex}-${colorIndex}`"
                        class="color-option layout horizontal center-vertical"
                        :class="{ 'is-selected': activeColor === color.title }"
                        @click="setColor(color, group.meta.name)"
                    >
                        <span style="flex-grow: 1">{{ color.title }}</span>
                        <div class="color-previews layout horizontal">
                            <div
                                class="color-preview"
                                :style="{
                                    backgroundColor: swapColorsInternally ? color.clrB : color.clrA,
                                    borderColor: getBorderColor(swapColorsInternally ? color.clrB : color.clrA)
                                }"
                            />
                            <div
                                class="color-preview"
                                :style="{
                                    backgroundColor: swapColorsInternally ? color.clrA : color.clrB,
                                    borderColor: getBorderColor(swapColorsInternally ? color.clrA : color.clrB)
                                }"
                            />
                        </div>
                    </div>
                </div>
            </div>
        </div>
        <div v-if="useCustomColor">
            <div class="layout horizontal m-t-8">
                <ipl-input
                    v-model="customColorA"
                    type="color"
                    style="flex-grow: 1"
                    name="team-a-color"
                />
                <ipl-input
                    v-model="customColorB"
                    type="color"
                    class="m-l-8"
                    style="flex-grow: 1"
                    name="team-b-color"
                />
            </div>
            <ipl-button
                label="Update"
                class="m-t-8"
                :color="customColorChanged ? 'red' : 'blue'"
                data-test="custom-color-submit-btn"
                @click="submitCustomColor"
            />
        </div>
    </ipl-expanding-space>
</template>

<script lang="ts">
import { computed, defineComponent, ref } from 'vue';
import { useActiveRoundStore } from '../../store/activeRoundStore';
import { getContrastingTextColor } from '../../helpers/colorHelper';
import { ColorInfo } from 'types/colors';
import { IplButton, IplExpandingSpace, IplCheckbox, IplInput } from '@iplsplatoon/vue-components';
import { themeColors } from '../../../browser/styles/colors';
import { useSettingsStore } from '../../settings/settingsStore';
import { perGameData } from '../../../helpers/gameData/gameData';

export default defineComponent({
    name: 'ColorEditor',

    components: { IplButton, IplInput, IplCheckbox, IplExpandingSpace },

    setup() {
        const activeRoundStore = useActiveRoundStore();
        const settingsStore = useSettingsStore();
        const gameData = computed(() => perGameData[settingsStore.state.runtimeConfig.gameVersion]);

        const useCustomColor = ref(false);
        const customColorA = ref(null);
        const customColorB = ref(null);

        activeRoundStore.watch(store => store.activeRound.teamA.color, newValue => {
            customColorA.value = newValue;
        }, { immediate: true });
        activeRoundStore.watch(store => store.activeRound.teamB.color, newValue => {
            customColorB.value = newValue;
        }, { immediate: true });

        activeRoundStore.watch(store => store.swapColorsInternally, () => {
            const colorA = customColorA.value;
            customColorA.value = customColorB.value;
            customColorB.value = colorA;
        });

        const swapColorsInternally = computed(() => activeRoundStore.state.swapColorsInternally);

        activeRoundStore.watch(store => store.activeRound.activeColor.isCustom, newValue => {
            useCustomColor.value = newValue;
        }, { immediate: true });

        function swapColors(data: ColorInfo): ColorInfo {
            return {
                ...data,
                clrA: data.clrB,
                clrB: data.clrA
            };
        }

        return {
            useCustomColor,
            customColorA,
            customColorB,
            customColorChanged: computed(() => {
                const activeRound = activeRoundStore.state.activeRound;
                return customColorA.value !== activeRound.teamA.color || customColorB.value !== activeRound.teamB.color;
            }),
            submitCustomColor() {
                activeRoundStore.dispatch('setActiveColor', {
                    categoryName: 'Custom Color',
                    color: {
                        index: 0,
                        title: 'Custom Color',
                        clrA: customColorA.value,
                        clrB: customColorB.value,
                        clrNeutral: '#FFFFFF',
                        isCustom: true
                    }
                });
            },
            colorsWithoutCustom: computed(() =>
                gameData.value.colors.filter(color => color.meta.name !== 'Custom Color')),
            activeColor: computed(() => activeRoundStore.state.activeRound.activeColor.title),
            getBorderColor(color: string): string {
                return getContrastingTextColor(color, 'white', themeColors.backgroundColorTertiary);
            },
            setColor(color: ColorInfo, categoryName: string): void {
                activeRoundStore.dispatch('setActiveColor', {
                    color: swapColorsInternally.value ? swapColors(color) : color,
                    categoryName
                });
            },
            swapColorsInternally
        };
    }
});
</script>

<style lang="scss" scoped>
@import '../../../browser/styles/colors';
@import '../../../browser/styles/constants';

.colors-container {
    max-height: 180px;
    overflow: auto;
}

.color-option {
    width: auto;
    margin: 8px 0;
    background-color: $background-secondary;
    border-radius: $border-radius-inner;
    padding: 6px 10px;
    cursor: pointer;
    transition-duration: $transition-duration-low;

    &.is-selected {
        background-color: $blue;

        &:hover {
            background-color: $blue-hover;
        }

        &:active {
            background-color: $blue-active;
        }
    }

    &:not(.is-selected) {
        &:hover {
            background-color: $background-secondary-hover;
        }

        &:active {
            background-color: $background-secondary-active;
        }
    }

    .color-previews {
        pointer-events: none;
    }

    .color-preview {
        width: 20px;
        border-style: solid;
        border-width: 2px;
        height: 20px;
        border-radius: 50%;
        margin: 1px;
        pointer-events: none;
        transition-duration: $transition-duration-med;
    }
}
</style>
