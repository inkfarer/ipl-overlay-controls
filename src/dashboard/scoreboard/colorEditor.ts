import { activeRound, swapColorsInternally } from './replicants';
import { ColorInfo } from 'types/colors';
import { fillColorSelector } from '../helpers/colorHelper';
import { getColorOptionName } from '../../helpers/splatoonData';
import { SetActiveColorRequest } from 'types/messages/activeRound';

const customColorToggle = <HTMLInputElement> document.getElementById('custom-color-toggle');
const colorSelector = <HTMLSelectElement> document.getElementById('color-selector');

fillColorSelector(colorSelector);

activeRound.on('change', newValue => {
    colorSelector.value = getColorOptionName(newValue.activeColor.index, newValue.activeColor.categoryName);

    document.getElementById('team-a-color-display').style.backgroundColor = newValue.teamA.color;
    document.getElementById('team-b-color-display').style.backgroundColor = newValue.teamB.color;
    (document.getElementById('team-a-custom-color') as HTMLInputElement).value = newValue.teamA.color;
    (document.getElementById('team-b-custom-color') as HTMLInputElement).value = newValue.teamB.color;

    const customColorEnabled = newValue.activeColor.isCustom;
    customColorToggle.checked = customColorEnabled;
    updateCustomColorToggle(customColorEnabled);
});

customColorToggle.addEventListener('change', e => {
    updateCustomColorToggle((e.target as HTMLInputElement).checked);
});

function updateCustomColorToggle(checked: boolean) {
    const colorSelectContainer = document.getElementById('color-select-container');
    const customColorContainer = document.getElementById('custom-color-select-container');

    if (checked) {
        colorSelectContainer.style.display = 'none';
        customColorContainer.style.display = 'flex';
    } else {
        colorSelectContainer.style.display = 'unset';
        customColorContainer.style.display = 'none';
    }
}

document.getElementById('swap-color-order-btn').addEventListener('click', () => {
    swapColorsInternally.value = !swapColorsInternally.value;
});

document.getElementById('update-scoreboard-btn').addEventListener('click', () => {
    const colorOption = <HTMLOptionElement> colorSelector.options[colorSelector.selectedIndex];

    let clrInfo: ColorInfo;
    const isCustomColor = customColorToggle.checked;

    if (isCustomColor) {
        clrInfo = {
            index: 0,
            title: 'Custom Color',
            clrA: (document.getElementById('team-a-custom-color') as HTMLInputElement).value,
            clrB: (document.getElementById('team-b-custom-color') as HTMLInputElement).value,
            isCustom: true
        };
    } else {
        clrInfo = {
            index: Number(colorOption.dataset.index),
            title: colorOption.text,
            clrA: swapColorsInternally.value ? colorOption.dataset.secondColor : colorOption.dataset.firstColor,
            clrB: swapColorsInternally.value ? colorOption.dataset.firstColor : colorOption.dataset.secondColor,
            isCustom: false
        };
    }

    nodecg.sendMessage('setActiveColor', {
        color: clrInfo,
        categoryName: isCustomColor ? 'Custom Color' : colorOption.dataset.categoryName
    } as SetActiveColorRequest);
});
