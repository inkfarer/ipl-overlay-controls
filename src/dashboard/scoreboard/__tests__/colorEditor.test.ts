import { MockNodecg } from '../../__mocks__/mockNodecg';
import { dispatch, elementById } from '../../helpers/elemHelper';

describe('colorEditor', () => {
    let nodecg: MockNodecg;

    beforeEach(() => {
        jest.resetModules();
        jest.resetAllMocks();

        nodecg = new MockNodecg();
        nodecg.init();

        document.body.innerHTML = `
            <select id="color-selector"></select>
            <input id="custom-color-toggle">
            <input id="team-a-custom-color">
            <input id="team-b-custom-color">
            <button id="swap-color-order-btn"></button>
            <button id="update-scoreboard-btn"></button>
            <div id="team-a-color-display"></div>
            <div id="team-b-color-display"></div>
            <div id="color-select-container"></div>
            <div id="custom-color-select-container"></div>
        `;

        require('../colorEditor');
    });

    it('fills color selector with options', () => {
        expect(elementById('color-selector').innerHTML).toMatchSnapshot();
    });

    describe('activeRound: change', () => {
        it('updates data displays', () => {
            nodecg.listeners.activeRound({
                activeColor: {
                    index: 1,
                    categoryName: 'Ranked Modes',
                    isCustom: false
                },
                teamA: { color: '#123123' },
                teamB: { color: '#FFFFFF' }
            });

            expect(elementById<HTMLSelectElement>('color-selector').value).toEqual('ranked-modes_1');
            expect(elementById('team-a-color-display').style.backgroundColor).toEqual('rgb(18, 49, 35)');
            expect(elementById('team-b-color-display').style.backgroundColor).toEqual('rgb(255, 255, 255)');
            expect(elementById<HTMLInputElement>('team-a-custom-color').value).toEqual('#123123');
            expect(elementById<HTMLInputElement>('team-b-custom-color').value).toEqual('#FFFFFF');
            expect(elementById<HTMLInputElement>('custom-color-toggle').checked).toEqual(false);
        });

        it('updates data displays if a custom color is given', () => {
            nodecg.listeners.activeRound({
                activeColor: {
                    index: 0,
                    categoryName: 'Custom Color',
                    isCustom: true
                },
                teamA: { color: '#123123' },
                teamB: { color: '#FFFFFF' }
            });

            expect(elementById<HTMLSelectElement>('color-selector').value).toEqual('custom-color_0');
            expect(elementById('team-a-color-display').style.backgroundColor).toEqual('rgb(18, 49, 35)');
            expect(elementById('team-b-color-display').style.backgroundColor).toEqual('rgb(255, 255, 255)');
            expect(elementById<HTMLInputElement>('team-a-custom-color').value).toEqual('#123123');
            expect(elementById<HTMLInputElement>('team-b-custom-color').value).toEqual('#FFFFFF');
            expect(elementById<HTMLInputElement>('custom-color-toggle').checked).toEqual(true);
            expect(elementById('color-select-container').style.display).toEqual('none');
            expect(elementById('custom-color-select-container').style.display).toEqual('flex');
        });
    });

    describe('custom-color-toggle: change', () => {
        it('displays custom color container when checked', () => {
            const toggle = elementById<HTMLInputElement>('custom-color-toggle');
            toggle.checked = true;

            dispatch(toggle, 'change');

            expect(elementById('color-select-container').style.display).toEqual('none');
            expect(elementById('custom-color-select-container').style.display).toEqual('flex');
        });

        it('hides custom color container when unchecked', () => {
            const toggle = elementById<HTMLInputElement>('custom-color-toggle');
            toggle.checked = false;

            dispatch(toggle, 'change');

            expect(elementById('color-select-container').style.display).toEqual('unset');
            expect(elementById('custom-color-select-container').style.display).toEqual('none');
        });
    });

    describe('swap-color-order-btn: click', () => {
        it('toggles swapColorsInternally value', () => {
            const button = elementById('swap-color-order-btn');
            nodecg.replicants.swapColorsInternally.value = true;

            dispatch(button, 'click');
            expect(nodecg.replicants.swapColorsInternally.value).toEqual(false);

            dispatch(button, 'click');
            expect(nodecg.replicants.swapColorsInternally.value).toEqual(true);
        });
    });

    describe('update-scoreboard-btn: click', () => {
        it('sends message to update color', () => {
            elementById<HTMLSelectElement>('color-selector').value = 'ranked-modes_1';

            dispatch(elementById('update-scoreboard-btn'), 'click');

            expect(nodecg.sendMessage).toHaveBeenCalledWith('setActiveColor', {
                color: {
                    index: 1,
                    title: 'Green vs Magenta',
                    clrA: '#04D976',
                    clrB: '#D600AB',
                    isCustom: false
                },
                categoryName: 'Ranked Modes'
            });
        });

        it('sends message to update color if color is custom', () => {
            elementById<HTMLSelectElement>('color-selector').value = 'ranked-modes_1';
            elementById<HTMLInputElement>('custom-color-toggle').checked = true;
            elementById<HTMLInputElement>('team-a-custom-color').value = '#123123';
            elementById<HTMLInputElement>('team-b-custom-color').value = '#567567';

            dispatch(elementById('update-scoreboard-btn'), 'click');

            expect(nodecg.sendMessage).toHaveBeenCalledWith('setActiveColor', {
                color: {
                    index: 0,
                    title: 'Custom Color',
                    clrA: '#123123',
                    clrB: '#567567',
                    isCustom: true
                },
                categoryName: 'Custom Color'
            });
        });
    });
});
