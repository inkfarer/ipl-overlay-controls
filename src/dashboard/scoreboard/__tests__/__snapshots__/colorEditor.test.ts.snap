// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`colorEditor fills color selector with options 1`] = `"<optgroup label=\\"Ranked Modes\\"><option value=\\"ranked-modes_0\\" data-index=\\"0\\" data-first-color=\\"#37FC00\\" data-second-color=\\"#7D28FC\\" data-category-name=\\"Ranked Modes\\">Green vs Grape</option><option value=\\"ranked-modes_1\\" data-index=\\"1\\" data-first-color=\\"#04D976\\" data-second-color=\\"#D600AB\\" data-category-name=\\"Ranked Modes\\">Green vs Magenta</option><option value=\\"ranked-modes_2\\" data-index=\\"2\\" data-first-color=\\"#10E38F\\" data-second-color=\\"#FB7B08\\" data-category-name=\\"Ranked Modes\\">Turquoise vs Orange</option><option value=\\"ranked-modes_3\\" data-index=\\"3\\" data-first-color=\\"#FF9E03\\" data-second-color=\\"#B909E0\\" data-category-name=\\"Ranked Modes\\">Mustard vs Purple</option><option value=\\"ranked-modes_4\\" data-index=\\"4\\" data-first-color=\\"#2F27CC\\" data-second-color=\\"#37FC00\\" data-category-name=\\"Ranked Modes\\">Dark Blue vs Green</option><option value=\\"ranked-modes_5\\" data-index=\\"5\\" data-first-color=\\"#B909E0\\" data-second-color=\\"#37FC00\\" data-category-name=\\"Ranked Modes\\">Purple vs Green</option><option value=\\"ranked-modes_6\\" data-index=\\"6\\" data-first-color=\\"#FEF232\\" data-second-color=\\"#2ED2FE\\" data-category-name=\\"Ranked Modes\\">Yellow vs Blue</option></optgroup><optgroup label=\\"Turf War\\"><option value=\\"turf-war_0\\" data-index=\\"0\\" data-first-color=\\"#D1E004\\" data-second-color=\\"#960CB2\\" data-category-name=\\"Turf War\\">Yellow vs Purple</option><option value=\\"turf-war_1\\" data-index=\\"1\\" data-first-color=\\"#E61077\\" data-second-color=\\"#361CB8\\" data-category-name=\\"Turf War\\">Pink vs Blue</option><option value=\\"turf-war_2\\" data-index=\\"2\\" data-first-color=\\"#ED0C6A\\" data-second-color=\\"#D5E802\\" data-category-name=\\"Turf War\\">Pink vs Yellow</option><option value=\\"turf-war_3\\" data-index=\\"3\\" data-first-color=\\"#6B10CC\\" data-second-color=\\"#08CC81\\" data-category-name=\\"Turf War\\">Purple vs Turquoise</option><option value=\\"turf-war_4\\" data-index=\\"4\\" data-first-color=\\"#E30960\\" data-second-color=\\"#02ADCF\\" data-category-name=\\"Turf War\\">Pink vs Light Blue</option><option value=\\"turf-war_5\\" data-index=\\"5\\" data-first-color=\\"#5617C2\\" data-second-color=\\"#FF5F03\\" data-category-name=\\"Turf War\\">Purple vs Orange</option><option value=\\"turf-war_6\\" data-index=\\"6\\" data-first-color=\\"#E60572\\" data-second-color=\\"#1BBF0F\\" data-category-name=\\"Turf War\\">Pink vs Green</option></optgroup><optgroup label=\\"Color Lock\\"><option value=\\"color-lock_0\\" data-index=\\"0\\" data-first-color=\\"#FEF232\\" data-second-color=\\"#2F27CC\\" data-category-name=\\"Color Lock\\">Yellow vs Blue (Color Lock)</option></optgroup><optgroup label=\\"Custom Color\\"><option value=\\"custom-color_0\\" data-index=\\"0\\" data-first-color=\\"#000000\\" data-second-color=\\"#FFFFFF\\" data-category-name=\\"Custom Color\\" disabled=\\"\\">Custom Color</option></optgroup>"`;
